### PR TITLE
[Feat][RayCluster] Use a new RayClusterReplicaFailure condition to re…

### PR DESCRIPTION
**Pull request by rueian**
…flect the result of creating/deleting Pods

## Checks

- [ ] I've made sure the tests are passing.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by kevin85421** [comment]
I think we should try to consolidate the code path to set up the condition. It is hard to maintain if we set the condition in multiple places.

**Comment by rueian** [comment]
> I think we should try to consolidate the code path to set up the condition. It is hard to maintain if we set the condition in multiple places.

Indeed, we should keep them in one place. Actually, I think the place should be the defer block in this draft and we will consolidate them in the https://github.com/ray-project/kuberay/issues/2235

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 1]
```diff
@@ -325,7 +333,7 @@ func (r *RayClusterReconciler) rayClusterReconcile(ctx context.Context, request
 	if calculateErr != nil {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:333]: I prefer not to set conditions in the reconcile functions. Instead, we should set conditions in `calculateStatus` based on `reconcileErr`.

```go
func calculateStatus(...) {
    if reconcileErr is PodError {
       // set condition
    }

}
```

**Review by kevin85421 — COMMENTED** [review]
* It is not very important for me to support fine-grained condition reasons. If we can easily support it, that is good. Otherwise, it is not necessary to support it, and use something like `PodFailure` instead.

* I prefer not to set conditions in the reconcile functions. Instead, we should set conditions in calculateStatus based on `reconcileErr`.
  ```go
  func calculateStatus(...) {
      if reconcileErr is PodError {
         // set condition
      }
  }
  ```

* I am not sure whether there is an easy way to determine whether an error is thrown by the Pod creation or deletion. If not, we can consider another way.

**Comment by rueian** [comment]
Hi @kevin85421,

> I prefer not to set conditions in the reconcile functions. Instead, we should set conditions in calculateStatus based on reconcileErr.

I am afraid that is not always feasible. We have multiple reconcileFuncs but have only one `reconcileErr`. Consider a case that the `reconcileErr` is not nil and is not an error from the `reconcilePods` function either, we will have difficulty with whether to turn the `RayClusterReplicaFailure` condition off or not. 

If the `reconcileErr` comes from other preceding reconcileFuncs, we should leave the condition untouched. 

If the `reconcileErr` comes from other succeeding reconcileFuncs, we should turn the condition off.

In other words, we will need to keep track of each execution of reconcileFuncs additionally if we set conditions based on the `reconcileErr`.

**Comment by kevin85421** [comment]
@rueian can we close this one?
